parquet/src/main/java/org/apache/iceberg/parquet/ParquetVariantVisitor.java

```diff
@@ -0,0 +1,287 @@
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName;
+import org.apache.parquet.schema.Type;
+
+public abstract class ParquetVariantVisitor<R> {
+  static final String METADATA = "metadata";
+  static final String VALUE = "value";
+  static final String TYPED_VALUE = "typed_value";
+
+  /**
+   * Handles the root variant column group.
+   *
+   * <p>The value and typed_value results are combined by calling {@link #value}.
+   *
+   * <pre>
+   *   group v (VARIANT) { &lt;-- metadata result and combined value and typed_value result
+   *     required binary metadata;
+   *     optional binary value;
+   *     optional ... typed_value;
+   *   }
+   * </pre>
+   */
+  public R variant(GroupType variant, R metadataResult, R valueResult) {
+    return null;
+  }
+
+  /**
+   * Handles a serialized variant metadata column.
+   *
+   * <pre>
+   *   group v (VARIANT) {
+   *     required binary metadata; &lt;-- this column
+   *     optional binary value;
+   *     optional ... typed_value;
+   *   }
+   * </pre>
+   */
+  public R metadata(PrimitiveType metadata) {
+    return null;
+  }
+
+  /**
+   * Handles a serialized variant value column.
+   *
+   * <pre>
+   *   group variant_value_pair {
+   *     optional binary value; &lt;-- this column
+   *     optional ... typed_value;
+   *   }
+   * </pre>
+   */
+  public R serialized(PrimitiveType value) {
+    return null;
+  }
+
+  /**
+   * Handles a shredded primitive typed_value column.
+   *
+   * <pre>
+   *   group variant_value_pair {
+   *     optional binary value;
+   *     optional int32 typed_value; &lt;-- this column when it is any primitive
+   *   }
+   * </pre>
+   */
+  public R primitive(PrimitiveType primitive) {
+    return null;
+  }
+
+  /**
+   * Handles a variant value result and typed_value result pair.
+   *
+   * <p>The value and typed_value pair may be nested in an object field, array element, or in the
+   * root group of a variant.
+   *
+   * <p>This method is also called when the typed_value field is missing.
+   *
+   * <pre>
+   *   group variant_value_pair { &lt;-- value result and typed_value result
+   *     optional binary value;
+   *     optional ... typed_value;
+   *   }
+   * </pre>
+   */
+  public R value(GroupType value, R valueResult, R typedResult) {
+    return null;
+  }
+
+  /**
+   * Handles a shredded object value result and a list of field value results.
+   *
+   * <p>Each field's value and typed_value results are combined by calling {@link #value}.
+   *
+   * <pre>
+   *   group variant_value_pair {  &lt;-- value result and typed_value field results
+   *     optional binary value;
+   *     optional group typed_value {
+   *       required group a {
+   *         optional binary value;
+   *         optional binary typed_value (UTF8);
+   *       }
+   *       ...
+   *     }
+   *   }
+   * </pre>
+   */
+  public R object(GroupType object, R valueResult, List<R> fieldResults) {
```

Comment:
The `value`, `object`, and `array` visitor methods are all used for a `value` / `typed_value` pair, depending on the structure of typed value (a shredded value, shredded array, or shredded object).